Signal/src/ViewControllers/MessageMetadataViewController.swift

```diff
@@ -0,0 +1,393 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+class MessageMetadataViewController: OWSViewController {
+
+    let TAG = "[MessageMetadataViewController]"
+
+    // MARK: Properties
+
+    let message: TSMessage
+
+    var mediaMessageView: MediaMessageView?
+
+    var scrollView: UIScrollView?
+    var contentView: UIView?
+
+    var dataSource: DataSource?
+    var attachmentStream: TSAttachmentStream?
+    var messageBody: String?
+
+    // MARK: Initializers
+
+    @available(*, unavailable, message:"use message: constructor instead.")
+    required init?(coder aDecoder: NSCoder) {
+        self.message = TSMessage()
+        super.init(coder: aDecoder)
+        owsFail("\(self.TAG) invalid constructor")
+    }
+
+    required init(message: TSMessage) {
+        self.message = message
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    // MARK: View Lifecycle
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        self.navigationItem.title = NSLocalizedString("MESSAGE_METADATA_VIEW_TITLE",
+                                                      comment: "Title for the 'message metadata' view.")
+
+        createViews()
+    }
+
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(animated)
+
+        mediaMessageView?.viewWillAppear(animated)
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+
+        mediaMessageView?.viewWillDisappear(animated)
+    }
+
+    // MARK: - Create Views
+
+    private func createViews() {
+        view.backgroundColor = UIColor.white
+
+        let scrollView = UIScrollView()
+        self.scrollView = scrollView
+        view.addSubview(scrollView)
+        scrollView.autoPinWidthToSuperview(withMargin:0)
+        scrollView.autoPin(toTopLayoutGuideOf: self, withInset:0)
+
+        let footer = UIToolbar()
+        footer.barTintColor = UIColor.ows_materialBlue()
+        view.addSubview(footer)
+        footer.autoPinWidthToSuperview(withMargin:0)
+        footer.autoPinEdge(.top, to:.bottom, of:scrollView)
+        footer.autoPin(toBottomLayoutGuideOf: self, withInset:0)
+
+        footer.items = [
+            UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil),
+            UIBarButtonItem(barButtonSystemItem: .action, target: self, action: #selector(shareButtonPressed)),
+            UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
+        ]
+
+        // See notes on how to use UIScrollView with iOS Auto Layout:
+        //
+        // https://developer.apple.com/library/content/releasenotes/General/RN-iOSSDK-6_0/
+        let contentView = UIView.container()
+        self.contentView = contentView
+        scrollView.addSubview(contentView)
+        contentView.autoPinLeadingToSuperView()
```

Comment:
`UIScrollView` vs. iOS AutoLayout is a strange relationship, since `UIScrollView`'s "content size" changes to reflect its subviews b) scrolls.  PureLayout and our `UIView+OWS` extensions to PureLayout are a nice abstraction for manipulating constraints, but given the "sensitivity" of this layout I wanted to use methods that translated 1:1 to constraints to be really explicit about the intent.